// three-msdf/webpack.config.js
const webpack = require('webpack')
const path = require('path')
const CopyWebpackPlugin = require('copy-webpack-plugin')

const paths = {
  src: path.join(__dirname, '/src'),
  dist: path.join(__dirname, '/build')
}

module.exports = {
  entry: {
    path: path.join(paths.src, 'index.js'),
  },
  output: {
    path: paths.dist,
    filename: 'three-msdf.js',
    library: "TextMSDF",
    libraryTarget: 'umd'
  },
  module: {
    rules: [
      {
        test: /\.(js)$/,
        loader: 'babel-loader',
        include: paths.src,
        query: {
          presets: ['es2015', 'stage-2']
        }
      },
      {
        test: /\.(glsl|frag|vert)$/,
        exclude: /node_modules/,
        use: [
          'raw-loader',
          'glslify'
        ]
      }
    ]
  },
  plugins: [
    new webpack.optimize.UglifyJsPlugin({
      compress: { warnings: false, screw_ie8: true, drop_console: true },
      output: { comments: false },
      mangle: { screw_ie8: true }
    }),
    new webpack.optimize.OccurrenceOrderPlugin()
  ],
}
