// three-msdf/src/utils/bmFontLoader.js
import loadFont from 'load-bmfont'

global.THREE = require('three')

import {
  ImageUtils,
  TextureLoader
} from 'three'

export default class BMFontLoader {
  /**
 * Constructor
 * @param  {Object} params
 */
  constructor(params) {
    this.value = params.value || null;
    this.asset = null;
    this.attributes = params.attributes || {};
  }

  /**
   * Load callback
   * @param  {Function} onComplete callback
   * @param  {Function} onError callback
   */
  load(onComplete, onError, onProgress) {
    this.onComplete = onComplete;
    this.onError = onError;
    this.onProgress = onProgress

    let _this = this

    loadFont(this.attributes.font, (err, font) => {
      if (err) throw err
      new TextureLoader().load(this.attributes.image,
        (texture) => {
          _this.asset = {
            font: font,
            texture: texture
          }
          _this.onComplete(_this.asset)
        }),
        onProgress,
        onError
    })
  }


	/**
	 * Return the asset
	 * @return {DOMNode} Dom Node of AssetItem
	 */
  getAsset() {
    return this.asset;
  }
}
